compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -753,8 +924,11 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MirUsedCollector<'a, 'tcx> {
                 target_ty,
             )
             | mir::Rvalue::Cast(mir::CastKind::DynStar, ref operand, target_ty) => {
-                let target_ty = self.monomorphize(target_ty);
                 let source_ty = operand.ty(self.body, self.tcx);
+                // *Before* monomorphizing, record that we already handled this mention.
+                self.used_mentioned_items
+                    .insert(MentionedItem::UnsizeCast { source_ty, target_ty });
```

Comment:
I haven't added the str/slice filter here. It's annoying to duplicate that code, and while mentioned_items get serialized into metadata, here it's entirely local to the already pretty expensive collection pass so it doesn't seem worth it. The `used_mentioned_items` optimization is effective as long as at least all the mentioned items are added to the set; it's okay if there are a few more.

Reply:
yea I think it's fine